### ISSUE TITLE
Generate function documentation block

### ISSUE THREAD

**Issue by mesqueeb** [juanfranblanco/vscode-solidity#102]
For JavaScript we all love JSDoc and I personally use it all the time.
For Solidity I find myself having to manually add documentation blocks above smart contract functions...

Will the Solidity plugin ever be able to click "Document this" to get a documentation block with all function params and necessary fields already added to the block?

Best!

**Comment by nharlow89** [comment]
I would like to see this functionality as well! JSDoc comment block generation does not work with this extension at the moment.

**Comment by obernardovieira** [comment]
It's a good idea. I would love it too.
Let's have some references if someone wants to start working on it.
* https://github.com/Microsoft/vscode-comment
* https://github.com/Microsoft/vscode/issues/3210

**Comment by juanfranblanco** [comment]
@obernardovieira  the vscode-comment sounds like a great and simple solution

**Comment by juanfranblanco** [comment]
Any thoughts on contributing directly to that extension or "copy" and modify slightly to match solidity?

**Comment by obernardovieira** [comment]
TBH, no, I just had a look around when I saw this issue. But I've been very busy these days. So I decided to put the links in case anyone wants to work on it. If nobody will touch it, I might do it.

**Comment by juanfranblanco** [comment]
Great, my question was around what will be best. I try to avoid copy and paste (lifting code) were possible. So I don't know if some slight modifications could be made and contribute to that one (although more complex).

Another option will be to integrate it here (copy and modify it), test it, and ask for integration to their main code base.

**Comment by obernardovieira** [comment]
Ohh, I see.
Well, with vscode-comment you have to select the line to comment, which does not happen with jsdoc itself. You just start writing the comment in the line above (like `/**`) and a suggestion for comment appears and generates de documentation, leaving only the variables description to feel, which can be written and then jump to the next just using TAB.

But getting to your question, you are still deciding either contribute in the other repository (extension) or add this kind feature is this repository (extension), right?

I would make it in this one. Non-solidity developers don't care about solidity comments feature in a base extension. What's your opinion?

**Comment by juanfranblanco** [comment]
@obernardovieira  yeah good point, it can be put here and it can be suggested later to them, as vscode-comment should be in theory (based on the name for other languages). Yep it works like xml comments in visual studio (big one).

**Comment by obernardovieira** [comment]
@juanfranblanco I started working on it and I decided to give a try on the https://github.com/Microsoft/vscode-comment. I realized that the *functionParser* might be too complex so I decided try with regular expressions and it might be very simples to parse. Look at this

Imagine I start with something like `function method(uint256 num, bytes32 val)` some regex like `function .*\((.*)\)` can give me back the paramenters, like `uint256 num, bytes32 b`. With this result I can parse them like `(uint256|bytes32)[ ](\w+)[ ]?,?[ ]?` andI can get some result like
**match 1**
 * *full match* - uint256 num, 
 * *group 1* - uint256
 * *group 2* - num

**match 2**
 * *full match* - bytes32 val
 * *group 1* - bytes32
 * *group 2* - val

Please, give me your thoughs on this, I might be doing it in the next coming days.

**Comment by rocky** [comment]
@obernardovieira  Here is one cool and reliable way this might be done. 

For 

```solidity 
function sell(uint256 numTokens) public {
      ....
}
```
        
After run through solc with the `--ast` option you will get in JSON (among a lot of other stuff which I have stripped out): 

```
            "name": "sell",
            "nodeType": "FunctionDefinition",
            "parameters": {
              "parameters": [
                {
                  "src": "526:17:1",
                  "name": "numTokens",
                  "typeDescriptions": {
                     "typeIdentifier": "t_uint256",
                     "typeString": "uint256"
                  },
              ],
              "src": "525:19:1"
            },
            "returnParameters": {
              "nodeType": "ParameterList",
              "parameters": [],
              "src": "552:0:1",
              "visibility": "public"
            },
      }
```

And code from [remix-lib](https://www.npmjs.com/package/remix-lib)  has some routines to assist navigating the AST (which I have modifed slightly so that it works on current as opposed to the "legacy") ast.  WIth this you'd just pick out what you want, reformat using [handlebars](https://www.npmjs.com/package/handlebars) and this would do it. 

You may think that adding a solc compile and adding that option is a bit of overhead. But here's the thing, all of that will be available if the MythX extension is included.  Including the libraries to parse this and the some handlebars code used for other purposes. 

Some weekend I may try a proof-of-concept code to show how this might be done and then maybe you can take that and polish it off.

**Comment by rocky** [comment]
@obernardovieira See https://github.com/rocky/solc-jsondoc for proof-of-concept code as a standalone program showing how this might work.

**Comment by obernardovieira** [comment]
Thanks @rocky, it's been a while. I had a problem. What about using OpenZeppelin in contracts? solc is giving me an error. Like `Error: Source "openzeppelin-solidity/contracts/ownership/Ownable.sol" not found: File outside of allowed directories.`
Thanks.

**Comment by rocky** [comment]
I am lost at what you are asking, and what exactly you tried.

Maybe if you give 

* the complete source code (or better stripped down as much as possible that still shows the problem), followed by 
* the command that you issued, 
* and the full output (if hte above is not complete),

then I might be able to better understand what the problem is. (By the way, this is a good thing to provide on all bug reports and issues). 

Oh, and also, do you use `solc` to compile your Solidity contracts? If not then how do your contracts get compiled?  Again, from my side it is better to err on giving too much information (which can be ignored if not needed) rather than too little.  Thanks.

**Comment by rocky** [comment]
I don't have special knowledge working with solc or of Solidty, but it looks like:

```
solc -o outDir --allow-paths=<path>/node_modules/openzeppelin-solidity/contracts ...
```

should have been 

```
solc -o outDir --allow-paths=<path>/node_modules ...
```

for an import that starts `openzeppelin-solidity/contracts`

**Comment by rocky** [comment]
Gee, I don't know.  A simple google search turns up https://github.com/ethereum/solidity/issues/2742

What does this have to do with the issue here? It seems to me that your question is about how to run solc.  So maybe https://ethereum.stackexchange.com/ ?

**Comment by juanfranblanco** [comment]
@obernardovieira @rocky mainly the compiler cannot resolve those paths, that is custom import resolving for open zeppelin or any other library format like dappsys. How is this related to the code generation of documentation? The output from solc (which won't succeed at the moment if there are syntax errors) should be picked up from the ide solc compiled output, that can be found in bin folder for testing. Although for the time being, until solc (solc.js) handles errors gracefully users would not want to compile their contract first and then generate the extra doc header, although I champion that approach for the time being it wont be a good UX to expect that.

**Comment by juanfranblanco** [comment]
On the other hand, i have just seen that sample from @rocky if it can be converted to read the solc.js standard output not the truffle one, that could be an option, I guess we could have a falloever code generation of doc headers if the compiler throws a wobbler.

**Comment by obernardovieira** [comment]
I got lost in the last message. And I did not really got the `we could have a falloever code generation of doc headers`

**Comment by juanfranblanco** [comment]
Oh mainly the solution uses solc(js) to generate a header/ documentation block. This relies on the contract to compile, as opposed to a normal ide which will generate the block without needing to be compiled successfuly. So an approach to do this will be to use solparse as it is currently used to generate autocomplete, navigation etc until solc can gracefully handle errors. 

On the other note, mainly is required the support of different library structures (i.e dappsys, openzeppelin, etc) as you have noted. Which is what is currently done with solparse.

**Comment by obernardovieira** [comment]
hmm ok,
I agree with "generate the block without needing to be compiled". In the beginning, the regex looked an awesome idea. I will have a look at **solparse** And I will say something soon. Thanks

**Comment by rocky** [comment]
In an experimental version, I now have function and contract documentation. A demo of this can be found at https://www.youtube.com/watch?v=SaaAYaEvrXE.

The [ ode itself for this](https://github.com/rocky/solc-vscode/blob/master/src/commands.ts#L149-L227) although it looks long is actually pretty simple and straightforward.

**Comment by LanfordCai** [comment]
Hi, how is it now?

**Comment by RickVM** [comment]
Would love to see this feature happen!

**Comment by ZrowGz** [comment]
Definitely having the ability to auto-generate a `.md` file from the natspec comments within contracts would be a super powerful tool & potentially help make creating adequate documentation a much easier process. Had there been any inroads with this?
